staging/src/k8s.io/apiserver/pkg/endpoints/discovery/aggregated/peer_aggregated_handler.go

```diff
@@ -0,0 +1,254 @@
+	serveHTTPFunc         func(http.ResponseWriter, *http.Request)
+}
+
+// NewPeerAggDiscoveryHandler creates a new handler for peer-aggregated discovery.
+func NewPeerAggDiscoveryHandler(localDiscoveryProvider ResourceManager, peerDiscoveryProvider PeerDiscoveryProvider, path string) PeerAggResourceManager {
+	scheme := runtime.NewScheme()
+	utilruntime.Must(apidiscoveryv2.AddToScheme(scheme))
+	codecs := serializer.NewCodecFactory(scheme)
+
+	h := &peerAggDiscoveryHandler{
+		localResourceManager:  localDiscoveryProvider,
+		peerDiscoveryProvider: peerDiscoveryProvider,
+		serializer:            codecs,
+	}
+
+	h.localResourceManager.AddInvalidationCallback(func() {
+		h.InvalidateCache()
+	})
+
+	// Instrumentation wrapper for serveHTTP
+	h.serveHTTPFunc = metrics.InstrumentHandlerFunc(request.MethodGet,
+		"", "", "", path, "", metrics.APIServerComponent, false, "",
+		func(resp http.ResponseWriter, req *http.Request) {
+			h.serveHTTP(resp, req)
+		})
+
+	return h
+}
+
+// InvalidateCache invalidates the peer-aggregated discovery caches.
+// This should be called when peer discovery data changes.
+func (h *peerAggDiscoveryHandler) InvalidateCache() {
+	h.cache.Store(nil)
+	klog.V(4).Info("Invalidated peer-aggregated discovery caches")
+}
+
+func (h *peerAggDiscoveryHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	h.serveHTTPFunc(resp, req)
+}
+
+func (h *peerAggDiscoveryHandler) serveHTTP(resp http.ResponseWriter, req *http.Request) {
+	cache := h.fetchFromCache()
+	response := cache.cachedResponse
+	etag := cache.cachedResponseETag
+
+	writeDiscoveryResponse(&response, etag, h.serializer, resp, req)
+}
+
+func (h *peerAggDiscoveryHandler) fetchFromCache() *cachedGroupList {
+	cacheLoad := h.cache.Load()
+	if cacheLoad != nil {
+		PeerAggCacheHitsCounter.Inc()
+		return cacheLoad
+	}
+
+	// Get local groups
+	var localGroups []apidiscoveryv2.APIGroupDiscovery
+	if rdm, ok := h.localResourceManager.(resourceManager); ok {
+		localCache := rdm.fetchFromCache()
+		if localCache != nil {
+			localGroups = localCache.cachedResponse.Items
+		}
+	}
+
+	// Get peer resources if provider is set
+	var mergedGroups []apidiscoveryv2.APIGroupDiscovery
+	if h.peerDiscoveryProvider != nil {
+		peerResources := h.peerDiscoveryProvider.GetPeerResources()
+		mergedGroups = h.mergeResources(localGroups, peerResources)
+	} else {
+		mergedGroups = localGroups
+	}
+
+	response := apidiscoveryv2.APIGroupDiscoveryList{
+		Items: mergedGroups,
+	}
+	etag, err := calculateETag(response)
+	if err != nil {
+		klog.Errorf("failed to calculate etag for discovery document: %v", err)
+		etag = ""
+	}
+
+	cached := &cachedGroupList{
+		cachedResponse:     response,
+		cachedResponseETag: etag,
+	}
+	h.cache.Store(cached)
+	return cached
+}
+
+func (h *peerAggDiscoveryHandler) mergeResources(
+	localGroups []apidiscoveryv2.APIGroupDiscovery,
+	peerGroupDiscovery map[string][]apidiscoveryv2.APIGroupDiscovery,
```

Comment:
how are peers ordered? what is the key of this map? I think we need to deterministically order our arguments to `SortDiscoveryGroupsTopo` to get consistent results, so accumulating the groups / versions we pass in by something like identity (including ordering our local values by our own local identity) might be a possibility

Reply:
key is the serverID. You're right, without passing in consistent ordered slices, we risk each server constructing a different ordering of the groups. I am now doing the following:
1. create a combined map from serverID -> []APIGroupDiscovery using both local and peer servers 
2. sort the keys of of this combined map and store them in a list
3. iterate over the sorted IDs list to process each server's []APIGroupDiscovery one by one